packages/core/strapi/lib/services/content-api/index.js

```diff
@@ -2,20 +2,36 @@
 
 const { uniq } = require('lodash');
 const permissions = require('./permissions');
+
+/**
+ * Creates an handler which check that the permission's action exists in the action registry
+ */
+const createValidatePermissionHandler =
+  (actionProvider) =>
+  ({ permission }) => {
+    const action = actionProvider.get(permission.action);
+
+    // If the action isn't registered into the action provider, then ignore the permission and warn the user
+    if (!action) {
+      strapi.log.debug(
+        `Unknown action "${permission.action}" supplied when registering a new permission`
+      );
+      return false;
+    }
+  };
+
 /**
  * Create a content API container that holds logic, tools and utils. (eg: permissions, ...)
  */
 const createContentAPI = (/* strapi */) => {
+  // Add providers
+  const providers = {
+    action: permissions.providers.createActionProvider(),
+    condition: permissions.providers.createConditionProvider(),
+  };
+
   const syncActions = async () => {
-    /**
-     * NOTE: For some reason, this doesn't seem to be necessary because all the routes exist
-     * createActionProvider uses a providerFactory, which seems to already include everything, and when we try
-     * to register our actions we get an error that the keys already exist
-     * Could providerFactory not be providing a new provider, and instead sharing the registry with everything that uses it?
-     *
-     * If this isn't an issue to fix and is expected, we don't need the route registration code below and it should be removed
-     * */
-    // Start of route registration
+    // Register actions
     const apiRoutesName = Object.values(strapi.api)
```

Comment:
I know this is what I initially pushed to create all the actions UID but I'm curious about whether we should use routes handler name (this) or controllers name (what we use in U&P for now).

The benefit I see by doing this approach is that we're not leaking controllers that are not associated with a route. But the downside is that if a controller is linked to a route dynamically, we'll miss it.

@alexandrebodin any opinion on this?

Reply:
ref: 

https://github.com/strapi/strapi/blob/main/packages/plugins/users-permissions/server/services/users-permissions.js#L154-L170

https://github.com/strapi/strapi/blob/main/packages/plugins/users-permissions/server/services/users-permissions.js#L32-L98